Sources/StreamChat/Repositories/AuthenticationRepository.swift

```diff
@@ -0,0 +1,190 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+public typealias TokenProvider = (@escaping (Result<Token, Error>) -> Void) -> Void
+
+class AuthenticationRepository {
+    private let tokenQueue: DispatchQueue = DispatchQueue(label: "io.getstream.auth-repository", attributes: .concurrent)
+    private var _isGettingToken: Bool = false
+    private var _currentUserId: UserId?
+    private var _currentToken: Token?
+    private var _tokenProvider: TokenProvider?
+    private var _tokenRequestCompletions: [(Error?) -> Void] = []
+
+    private var isGettingToken: Bool {
+        get { tokenQueue.sync { _isGettingToken } }
+        set { tokenQueue.async(flags: .barrier) { self._isGettingToken = newValue }}
+    }
+
+    private(set) var currentUserId: UserId? {
+        get { tokenQueue.sync { _currentUserId } }
+        set { tokenQueue.async(flags: .barrier) { self._currentUserId = newValue }}
+    }
+
+    private(set) var currentToken: Token? {
+        get { tokenQueue.sync { _currentToken } }
+        set { tokenQueue.async(flags: .barrier) {
+            self._currentToken = newValue
+            newValue.map { self._currentUserId = $0.userId }
+        }}
+    }
+
+    private(set) var tokenProvider: TokenProvider? {
+        get { tokenQueue.sync { _tokenProvider } }
+        set { tokenQueue.async(flags: .barrier) { self._tokenProvider = newValue }}
+    }
+
+    private(set) var tokenRequestCompletions: [(Error?) -> Void] {
+        get { tokenQueue.sync { _tokenRequestCompletions } }
+        set { tokenQueue.async(flags: .barrier) { self._tokenRequestCompletions = newValue }}
+    }
+
+    private let apiClient: APIClient
+    private let databaseContainer: DatabaseContainer
+    private let clientUpdater: ChatClientUpdater
+    /// A timer that runs token refreshing job
+    private var tokenRetryTimer: TimerControl?
+    /// Retry timing strategy for refreshing an expiried token
+    private var tokenExpirationRetryStrategy: RetryStrategy
+    private let timerType: Timer.Type
```

Comment:
Keep in mind we have a new `ScheduledStreamTimer: StreamTimer` that might be better than this one, especially for testing

Reply:
Stuff added here: https://stream-io.atlassian.net/browse/CIS-2233